content/blog/2025-04-27-valkey-modules-rust-sdk-updates.md

```diff
@@ -0,0 +1,272 @@
+
+```rust
+fn preload_fn(ctx: &Context, _args: &[ValkeyString]) -> Status {
+    let _version = ctx.get_server_version().unwrap();
+    // respond with Status::Ok or Status::Err to prevent loading
+    Status::Ok
+}
+valkey_module! {
+    ...
+    preload: preload_fn,
+    commands: [],
+}
+```
+
+## Filters
+
+To execute custom code before specific Valkey commands we can use command filters.  This can be leveraged to replace default commands with custom comands or modify arguments.
+Thanks to the abstractions provided by the SDK we simply need to create a Rust function and register it in the `valkey_module!` macro.
+Note of caution - since filters are executed before every command this code needs to be optimized.
+
+```rust
+fn my_filter_fn(_ctx: *mut RedisModuleCommandFilterCtx) {
+    let cf_ctx = CommandFilterCtx::new(ctx);
+    // check the number of arguments
+    if cf_ctx.args_count() != 3 {
+        return;
+    }
+    // get which command is being executed
+    let _cmd = cf_ctx.cmd_get_try_as_str().unwrap();
+    // grab various args passed to the command
+    let _all_args = cf_ctx.get_all_args_wo_cmd();
+    // replace command
+    cf_ctx.arg_replace(0, "custom_cmd");
+}
+valkey_module! {
+    ...
+    filters: [
+        [my_filter_fn, VALKEYMODULE_CMDFILTER_NOSELF],
+    ]
+}
+```
+
+## New event handlers
+
+SDK now supports more server events.  We can use this to execute our own code on client connect/disconnect, server shutdown or specific key events.
+
+```rust
+#[client_changed_event_handler]
+fn client_changed_event_handler(_ctx: &Context, client_event: ClientChangeSubevent) {
+    match client_event {
+        ClientChangeSubevent::Connected => {}
+        ClientChangeSubevent::Disconnected => {}
+    }
+}
+#[shutdown_event_handler]
+fn shutdown_event_handler(_ctx: &Context, _event: u64) {
+    ...
+}
+#[key_event_handler]
+fn key_event_handler(ctx: &Context, key_event: KeyChangeSubevent) {
+    match key_event {
+        KeyChangeSubevent::Deleted => {}
+        KeyChangeSubevent::Evicted => {}
+        KeyChangeSubevent::Overwritten => {}
+        KeyChangeSubevent::Expired => {}
+    }
+}
+```
+
+## Custom ACL categories support
+
+Valkey 8 introduced support for custom ACL categories.  To implement that we need to enable `required-features = ["min-valkey-compatibility-version-8-0"]` in `Cargo.toml` and register new categories in `valkey_module!` macro.  Then we can restrict our custom commands to these ACL categories.
+
+```rust
+valkey_module! {
+    ...
+    acl_categories: [
+        "acl_one",
+        "acl_two"
+    ]
+    commands: [
+        ["cmd1", cmd1, "write",  0, 0, 0, "acl_one"],
+        ["cmd2", cmd2, "", 0, 0, 0, "acl_one acl_two"],
+    ]
+```
+
+## Validating / Rejecting CONFIG SET
+
+SDK now supports specifying optional callback functions to validate or reject custom configuations.  This is available for String, i64, bool and enum configs.  Here is an example of validation for i64 custom config.  
```

Comment:
Definite article: 'SDK now supports specifying optional callback...' -> '**The** SDK now supports specifying optional callback'

Types should be in back ticks to make them inline code.

Reply:
updated